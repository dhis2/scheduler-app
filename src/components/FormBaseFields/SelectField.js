import React from 'react'
import { node, bool, shape, string, object } from 'prop-types'
import { SelectField as CoreSelect, Help } from '@dhis2/ui-core'

const SelectField = ({ input, meta, label, required, children }) => {
    const { touched, error } = meta
    const hasError = touched && !!error

    return (
        <React.Fragment>
            <CoreSelect
                {...input}
                label={label}
                error={hasError}
                required={required}
            >
                {children}
            </CoreSelect>
            {hasError && <Help error>{meta.error.message}</Help>}
        </React.Fragment>
    )
}

SelectField.defaultProps = {
    required: false,
}

SelectField.propTypes = {
    label: string.isRequired,
    meta: shape({
        error: object,
        touched: bool.isRequired,
    }).isRequired,
    input: object.isRequired,
    children: node.isRequired,
    required: bool,
}

export default SelectField
