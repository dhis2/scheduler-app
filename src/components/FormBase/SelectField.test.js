import React from 'react'
import { shallow } from 'enzyme'
import { Help } from '@dhis2/ui-core'
import SelectField from './SelectField'

describe('<SelectField>', () => {
    it('renders correctly', () => {
        const props = {
            input: {
                value: 'value',
                onChange: () => {},
                name: 'name',
            },
            meta: {
                touched: false,
            },
            label: 'label',
        }
        const wrapper = shallow(
            <SelectField {...props}>
                <option value="option">option</option>
            </SelectField>
        )

        expect(wrapper).toMatchSnapshot()
    })

    it('renders errors correctly', () => {
        const props = {
            input: {
                value: 'value',
                onChange: () => {},
                name: 'name',
            },
            meta: {
                touched: true,
                error: {
                    message: 'error',
                },
            },
            label: 'label',
        }
        const wrapper = shallow(
            <SelectField {...props}>
                <option value="option">option</option>
            </SelectField>
        )

        expect(wrapper.find(Help)).toHaveLength(1)
        expect(wrapper).toMatchSnapshot()
    })

    it('does not render errors if the field has not been touched', () => {
        const props = {
            input: {
                value: 'value',
                onChange: () => {},
                name: 'name',
            },
            meta: {
                touched: false,
                error: {
                    message: 'error',
                },
            },
            label: 'label',
        }
        const wrapper = shallow(
            <SelectField {...props}>
                <option value="option">option</option>
            </SelectField>
        )

        expect(wrapper.find(Help)).toHaveLength(0)
    })
})
