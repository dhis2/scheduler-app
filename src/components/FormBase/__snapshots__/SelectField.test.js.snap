// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectField> renders correctly 1`] = `
<Fragment>
  <SelectField
    error={false}
    label="label"
    name="name"
    onChange={[Function]}
    required={false}
    value="value"
  >
    <option
      value="option"
    >
      option
    </option>
  </SelectField>
</Fragment>
`;

exports[`<SelectField> renders errors correctly 1`] = `
<Fragment>
  <SelectField
    error={true}
    label="label"
    name="name"
    onChange={[Function]}
    required={false}
    value="value"
  >
    <option
      value="option"
    >
      option
    </option>
  </SelectField>
  <Help
    error={true}
  >
    error
  </Help>
</Fragment>
`;
